x-pack/plugins/ingest_manager/common/types/models/datasource.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { schema, TypeOf } from '@kbn/config-schema';
+
+export enum InputType {
+  Etc = 'etc',
+  Log = 'log',
+  MetricDocker = 'metric/docker',
+  MetricSystem = 'metric/system',
+}
+
+const DatasourceBaseSchema = {
+  name: schema.string(),
+  namespace: schema.maybe(schema.string()),
+  read_alias: schema.maybe(schema.string()),
+  agent_config_id: schema.string(),
+  package: schema.maybe(
+    schema.object({
+      assets: schema.arrayOf(
+        schema.object({
+          id: schema.string(),
+          type: schema.string(),
+        })
+      ),
+      description: schema.string(),
+      name: schema.string(),
+      title: schema.string(),
+      version: schema.string(),
+    })
+  ),
+  streams: schema.arrayOf(
+    schema.object({
+      config: schema.recordOf(schema.string(), schema.any()),
+      input: schema.object({
+        type: schema.oneOf([
+          schema.literal(InputType.Etc),
+          schema.literal(InputType.Log),
+          schema.literal(InputType.MetricDocker),
+          schema.literal(InputType.MetricSystem),
+        ]),
+        config: schema.recordOf(schema.string(), schema.any()),
+        fields: schema.maybe(schema.arrayOf(schema.recordOf(schema.string(), schema.any()))),
+        ilm_policy: schema.maybe(schema.string()),
+        index_template: schema.maybe(schema.string()),
+        ingest_pipelines: schema.maybe(schema.arrayOf(schema.string())),
+      }),
+      output_id: schema.string(),
+      processors: schema.maybe(schema.arrayOf(schema.string())),
+    })
+  ),
+};
+
+export const NewDatasourceSchema = schema.object({
+  ...DatasourceBaseSchema,
+});
+
+export const DatasourceSchema = schema.object({
+  ...DatasourceBaseSchema,
+  id: schema.string(),
+});
+
+export type NewDatasource = TypeOf<typeof NewDatasourceSchema>;
+
+export type Datasource = TypeOf<typeof DatasourceSchema>;
```

Comment:
This is a nifty shorthand, but I think this might create a conflict for naming conventions on the type.  Schema uses `output_id` but accessing in code should be `datasource.outputId` and I think it won't be.
<img width="504" alt="Screen Shot 2020-02-12 at 11 07 45 AM" src="https://user-images.githubusercontent.com/16563603/74353523-07cd2f80-4d88-11ea-9236-4fe397a1d5fa.png">

Reply:
@stacey-gammon, @jen-huang and I also talked about this with @joshdover. We'll leave it as-is for now and come back to it if the mismatch between property casing presents a problem. For now, we think the benefits of writing the config schema and easily extracting that to TS types outweigh the drawbacks.